File: compress-illus.py
"""
成知成语卡 - 插图压缩脚本（纯净版）
策略：最长边>1600px 则等比缩小，统一用 JPEG q=82 重存，去除 EXIF
只覆盖真正变小的文件
"""

import os
import sys
import shutil
import time

ILLUS_DIR = r"D:\Desktop\成知应用开发\2026-05-14-task-5\v8\illus"
JPEG_QUALITY = 82
MAX_SIZE = 1600
SUPPORTED = ('.jpg', '.jpeg', '.png', '.webp', '.bmp')

def compress_one(fname):
    path = os.path.join(ILLUS_DIR, fname)
    ext = os.path.splitext(fname)[1].lower()
    if ext not in SUPPORTED:
        return None, None, False

    from PIL import Image
    try:
        img = Image.open(path)
        # 转 RGB（处理 RGBA/P 模式）
        if img.mode in ('RGBA', 'P', 'LA'):
            bg = Image.new('RGB', img.size, (255, 255, 255))
            if img.mode == 'P':
                img = img.convert('RGBA')
            bg.paste(img, mask=img.split()[-1] if img.mode == 'RGBA' else None)
            img = bg
        elif img.mode != 'RGB':
            img = img.convert('RGB')

        # 等比缩小
        w, h = img.size
        if max(w, h) > MAX_SIZE:
            scale = MAX_SIZE / max(w, h)
            nw, nh = int(w * scale), int(h * scale)
            img = img.resize((nw, nh), Image.LANCZOS)

        # 先存到脚本目录的临时文件（避免同目录锁冲突）
        tmp = os.path.join(os.path.dirname(os.path.abspath(__file__)), fname + '.tmp.jpg')
        img.save(tmp, 'JPEG', quality=JPEG_QUALITY, optimize=True, progressive=True)
        img.close()

        orig_size = os.path.getsize(path)
        new_size = os.path.getsize(tmp)

        if new_size < orig_size:
            # 尝试覆盖，若被锁则重试3次
            for _attempt in range(3):
                try:
                    os.replace(tmp, path)
                    return orig_size, new_size, True
                except PermissionError:
                    time.sleep(0.5)
            # 仍失败，回退用 shutil.copyfile
            try:
                shutil.copyfile(tmp, path)
                os.remove(tmp)
                return orig_size, new_size, True
            except Exception:
                os.remove(tmp)
                return orig_size, orig_size, False
        else:
            os.remove(tmp)
            return orig_size, orig_size, False
    except Exception as e:
        # 清理可能的临时文件
        for t in (
            os.path.join(os.path.dirname(os.path.abspath(__file__)), fname + '.tmp.jpg'),
            path + '.tmp.jpg'
        ):
            try:
                if os.path.exists(t): os.remove(t)
            except:
                pass
        return None, None, False

def main():
    files = sorted([
        f for f in os.listdir(ILLUS_DIR)
        if os.path.splitext(f)[1].lower() in SUPPORTED
    ])

    print(f"共找到 {len(files)} 张图片，开始压缩...")
    total_before = 0
    total_after = 0
    saved = 0
    skipped = 0
    errors = 0

    for i, fname in enumerate(files, 1):
        before, after, ok = compress_one(fname)
        if before is None:
            errors += 1
            print(f"[{i}/{len(files)}] ERR  {fname}")
            continue

        pct = (after / before * 100) if before > 0 else 100
        if ok:
            saved += 1
            total_before += before
            total_after += after
            tag = "SAVED" if before > 10240 else "small"
        else:
            skipped += 1
            tag = "skip "

        bar = f"[{i}/{len(files)}] {tag}  {fname[:28]:<30} {before/1024:.1f}KB -> {after/1024:.1f}KB ({pct:.1f}%)"
        print(bar)

    print("\n" + "="*50)
    if total_before > 0:
        tot_pct = total_after / total_before * 100
        print(f"压缩完成：{saved} 张压缩，{skipped} 张跳过，{errors} 张失败")
        print(f"总大小：{total_before/1048576:.1f} MB -> {total_after/1048576:.1f} MB ({tot_pct:.1f}%)")
        print(f"节省空间：{(total_before-total_after)/1048576:.1f} MB")
    else:
        print(f"无需压缩的文件：{skipped} 张，错误：{errors} 张")

if __name__ == '__main__':
    main()

File: compress-illus.js
/**
 * 成知·成语卡 — 插图批量压缩脚本
 *
 * 用途：
 *   1. 等比缩放：宽或高 > 1600px 的图，等比缩小到最长边 = 1600px
 *   2. JPEG 重压缩：质量 85，去掉 EXIF / 元数据
 *   3. 原地替换：直接用压缩后的文件覆盖原文件
 *   4. 汇报结果：每张图的压缩比、总节省空间
 *
 * 用法：
 *   node compress-illus.js
 *
 * 依赖：
 *   npm install sharp   （先在项目的 Node 环境里安装）
 *
 * 注意：
 *   - 建议先在 illus/ 的副本里试跑，确认画质可接受再覆盖原文件
 *   - sharp 比 Pillow 更快，且对 JPEG 的压缩质量更好
 */

const fs   = require('fs');
const path = require('path');
const sharp = require('sharp');

const ILLUS_DIR   = path.join(__dirname, 'illus');
const MAX_DIM     = 1600;          // 最长边上限（px）
const JPEG_QUALITY = 85;            // JPEG 质量（0-100，85 肉眼无损）

async function compressOne(file) {
  const src = path.join(ILLUS_DIR, file);
  const tmp = src + '.tmp.jpg';

  const stBefore = fs.statSync(src);
  const beforeKB = (stBefore.size / 1024).toFixed(1);

  let pipeline = sharp(src, { failOnError: false });

  // 获取尺寸，决定是否缩放
  const meta = await pipeline.metadata();
  const { width, height } = meta;

  if (width > MAX_DIM || height > MAX_DIM) {
    pipeline = pipeline.resize({
      width:  width  > height ? MAX_DIM : undefined,
      height: height > width  ? MAX_DIM : undefined,
      fit: 'inside',
      withoutEnlargement: true,
    });
  }

  await pipeline
    .jpeg({ quality: JPEG_QUALITY, progressive: true, mozjpeg: true, strip: true })
    .toFile(tmp);

  const stAfter = fs.statSync(tmp);
  const afterKB = (stAfter.size / 1024).toFixed(1);
  const ratio   = (stAfter.size / stBefore.size * 100).toFixed(1);

  // 只有真的小了才替换
  if (stAfter.size < stBefore.size) {
    fs.renameSync(tmp, src);
    return { file, beforeKB, afterKB, ratio, saved: true };
  } else {
    fs.unlinkSync(tmp);   // 没变小，删掉临时文件
    return { file, beforeKB, beforeKB, ratio: '100.0', saved: false };
  }
}

(async () => {
  const files = fs.readdirSync(ILLUS_DIR)
    .filter(f => /\.(jpe?g|png)$/i.test(f));

  console.log(`共 ${files.length} 张图片，开始压缩…\n`);

  let totalBefore = 0;
  let totalAfter  = 0;
  let done = 0;

  for (const file of files) {
    done++;
    const r = await compressOne(file);
    totalBefore += parseFloat(r.beforeKB);
    if (r.saved) totalAfter += parseFloat(r.afterKB);

    const flag = r.saved ? '✅' : '⚠️ 未变小';
    console.log(`[${done}/${files.length}] ${flag}  ${r.file.padEnd(35)}  ${r.beforeKB.padStart(8)} KB → ${r.afterKB.padStart(8)} KB  (${r.ratio}%)`);
  }

  console.log(`\n✅ 完成！`);
  console.log(`压缩前总计：${(totalBefore/1024).toFixed(1)} MB`);
  console.log(`压缩后总计：${(totalAfter /1024).toFixed(1)} MB`);
  console.log(`节省空间：  ${((totalBefore - totalAfter)/1024).toFixed(1)} MB`);
})();
